### PR TITLE
record encryption

### PR DESCRIPTION
**Pull request by conradludgate**
Using a combination of V4 PASETO/PASERK with randomized wrapped keys to encrypt the data of each record.

I've used a set of `DecryptedData`/`EncryptedData` types to ensure that we don't mistakenly write decrypted data to the database

TODO:
- [x] record tag+version -> PASETO version
- [x] consider migrating the database to use a string for the data row instead of bytes

### CODE REVIEW

File: atuin-client/src/record/encryption.rs [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+This would be a lot more secure than using the master key directly.
+
+However, there's an established norm of using a random key. This scheme might be otherwise known as
+- client-side encryption
+- envelope encryption
+- key wrapping
+
+A HSM (Hardward security module) provider, eg AWS, Azure, GCP, or even a physical device like a yubikey
+will have some keys that they keep to themselves. These keys never leave their physical hardware.
+If they never leave the hardward, then encrypting large amounts of data means giving them the data and waiting.
```
Comment (mercxry) [inline atuin-client/src/record/encryption.rs:29]: Small typo ;)
```suggestion
If they never leave the hardware, then encrypting large amounts of data means giving them the data and waiting.
```

File: atuin-client/src/record/encryption.rs [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+Why do we use a random content-encryption key?
+Originally I was planning on using a derived key for encryption based on additional data.
+This would be a lot more secure than using the master key directly.
+
+However, there's an established norm of using a random key. This scheme might be otherwise known as
+- client-side encryption
+- envelope encryption
+- key wrapping
+
+A HSM (Hardward security module) provider, eg AWS, Azure, GCP, or even a physical device like a yubikey
```
Comment (mercxry) [inline atuin-client/src/record/encryption.rs:27]: Wouldn't wanna leave for the Hacktoberfest 🤣
```suggestion
A HSM (Hardware Security Module) provider, eg: AWS, Azure, GCP, or even a physical device like a YubiKey
```

File: atuin-common/src/record.rs [hunk 3]
```diff
@@ -35,17 +45,26 @@ pub struct Record {
     pub tag: String,
 
     /// Some data. This can be anything you wish to store. Use the tag field to know how to handle it.
-    pub data: Vec<u8>,
+    pub data: Data,
+}
+
+/// Extra data from the record that should be encoded in the data
+#[derive(Debug, Copy, Clone)]
+pub struct AdditonalData<'a> {
```
Comment (ellie) [inline atuin-common/src/record.rs:53]:
```suggestion
pub struct AdditionalData<'a> {
```

sorry 😭

**Review by ellie — CHANGES_REQUESTED** [review]
This is great, thank you! Just the spelling error and we're good

**Review by ellie — APPROVED** [review]

**Pull request merged**